pxr/usd/plugin/usdDraco/importTranslator.cpp

```diff
@@ -0,0 +1,272 @@
+            const draco::PointIndex pi = _dracoMesh.CornerToPointId(co);
+            isNewEdge = _addedEdges.GetMappedValue(pi);
+        }
+        if (isNewEdge) {
+            // Visit triangle across the new edge.
+            const draco::FaceIndex oppositeFaceIndex = cornerTable->Face(co);
+            _FindOriginalFaceEdges(oppositeFaceIndex, cornerTable,
+                                   triangleVisited, polygonEdges);
+        } else {
+            // Insert the original edge to the map.
+            const draco::PointIndex pointFrom = face[(c + 1) % 3];
+            draco::PointIndex pointTo = face[(c + 2) % 3];
+            polygonEdges.insert(
+                {PositionIndex(_positions.GetMappedIndex(pointFrom)),
+                 pointTo});
+        }
+    }
+}
+
+SdfLayerRefPtr UsdDracoImportTranslator::_Translate()
+{
+    // Create USD layer and stage.
+    SdfLayerRefPtr layer = SdfLayer::CreateAnonymous(".usda");
+    UsdStageRefPtr stage = UsdStage::Open(layer);
+
+    // Get data from Draco mesh.
+    _PopulateValuesFromMesh();
+
+    // Check validity of attributes.
+    if (!_CheckData())
+        return layer;
+
+    // Populate indices.
+    _PopulateIndicesFromMesh();
+
+    // Create USD mesh and set attributes.
+    UsdGeomMesh usdMesh = UsdGeomMesh::Define(stage, SdfPath("/DracoMesh"));
+    _SetAttributesToMesh(&usdMesh);
+
+    // Set the mesh as the default prim of the stage.
+    stage->SetDefaultPrim(usdMesh.GetPrim());
+    return layer;
+}
+
+void UsdDracoImportTranslator::_PopulateValuesFromMesh() {
+    if (_posOrder.HasPointAttribute()) {
+        const size_t numFaces = _dracoMesh.num_faces();
+        _positions.PopulateValuesWithOrder(_posOrder, numFaces, _dracoMesh);
+    } else {
+        _positions.PopulateValues();
+    }
+    _texCoords.PopulateValues();
+    _normals.PopulateValues();
+}
+
+void UsdDracoImportTranslator::_PopulateIndicesFromMesh() {
+    // Allocate index arrays as if all faces were triangles. The arrays will be
+    // downsized if the mesh turns out to have quads and other polygons.
+    const size_t numFaces = _dracoMesh.num_faces();
+    const size_t numCorners = 3 * numFaces;
+    _faceVertexCounts.resize(numFaces);
+    _faceVertexIndices.resize(numCorners);
+    _texCoords.ResizeIndices(numCorners);
+    _normals.ResizeIndices(numCorners);
+
+    // Create corner table.
+    std::unique_ptr<draco::CornerTable> cornerTable =
+        draco::CreateCornerTableFromPositionAttribute(&_dracoMesh);
+
+    // Reconstruct polygons here.
+    size_t vertexIndex = 0;
+    size_t faceIndex = 0;
+    std::vector<bool> triangleVisited(numFaces, false);
+
+    // Populate index arrays.
+    for (size_t i = 0; i < numFaces; i++) {
+        const draco::Mesh::Face &face = _dracoMesh.face(draco::FaceIndex(i));
+        if (_addedEdges.HasPointAttribute()) {
+            draco::FaceIndex fi(i);
+            PolygonEdges polygonEdges;
```

Comment:
PolygonEdges is a map so it would have to be dynamically allocated / deallocated for every face which is very inefficient. I would declare it before the for loop and clear it here.

Reply:
Done.